app/External/Slack/Modals/EquipmentAuthorization.php

```diff
@@ -96,26 +96,19 @@ public static function handle(SlackRequest $request)
         $selectedEquipment = self::equipmentFromState($state);
         $selectedMembers = self::peopleFromState($state);
```

Comment:
I see no place that uses anything about the customer here but their id. I think this is my fault but it just became more noticeable since we're doing N database queries for it. If you absolutely do need every Customer object, pass an array/collection of ids into the find method of customer to do just one query.

Reply:
The usage for the customer object is `customer->hasMembership` in order to skip redundant permissions, which is now handled in the actions. We could skip that check and instead rely on the operation being idempotent, or we could make this more efficient. I'm happy to batch queries for efficiency, but if four members is causing enough of a performance hit to be "more noticeable" then we have serious performance issues we need to get after. At the traffic volume we should be handling with our member base, database performance should not be an issue.

Reply:
The `hasMembership` call was what I missed and definitely means we should be casting to the customer and not the id. So what I would say is to just change your peopleFromState and equipmentFromState to call find on the mapped collection to get it down to 2 database queries. I see no reason why we should leave <number of equipment> + <number of persons> queries when we can easily get it down to 2 with no penalty.

Let's break it down a bit. As far as "more noticeable", we're seeing two effects at play. The first is what this PR wants to fix. The Slack modal fails because we try to make a bunch of requests at the same time which causes the user to hit submit again. The hasMembership check is only on the memberships that we've received updates for and we probably haven't received any of them yet. So when we submit again, some of the new memberships will get a 400 response with this redacted body:
```json
{
	"code": "woocommerce_rest_wc_user_membership_exists",
	"message": "A membership already exists for user <user id> and plan <plan id>.",
	"data": {
		"status": 400
	}
}
```

I would probably have made an Action for `CreateUserMembership` instead of using the business logic name of `AuthorizeEquipmentAction` (or `AuthorizeEquipmentUse`) just in case that needs to be re-used elsewhere. That way we can handle the 400 error in one location and check for the `code` field being `woocommerce_rest_wc_user_membership_exists`. But that's not really too much of an issue here, because we won't see the user submitting it multiple times now since our Slack modal will return a response in time.

The second issue is that [conversation.list](https://api.slack.com/methods/conversations.list) is a Tier 2 api rate limit, meaning it can have at most 20 per minute. We actually get both a created and updated webhook event for every user membership (well for every model really, if you get a create, you get an update and they are NOT guaranteed to be in order). If 4 pieces of equipment are being trained, with 3 members being trained on each, that's 24 times our AddToSlackChannel gets called. I'm currently working on fixing that side of it, specifically to handle the rate limiting better. I wouldn't call that "serious issues".

Reply:
> The hasMembership check is only on the memberships that we've received updates for and we probably haven't received any of them yet.

That's not actually why I do that check. I do this check because there are (somewhat routinely) real world scenarios where a subset of the members are authorized on a subset of the machines. Having attended the full class, it is convenient for the trainer to "reauthorize" these users by including them in the bulk submission for the class even though they already are signed off on some of the equipment - rather than submitting a separate form with only the equipment they need.

> wouldn't call that "serious issues"

I didn't say anything about the rate limit problem. You mentioned that it was "more noticeable" without saying what that was. Since you were calling out query efficiency, I made the deduction that you _might_ be referring to database performance. If so, that would indicate that something is really wrong. If there's no evidence of database slowness, great! I agree that we should make the query more efficient nonetheless.

Reply:
> That's not actually why I do that check.
I understand why you do the check, and it's a good check to do, I was just pointing out that given the current situation, when the user sees an error, it gets lovingly compounded if they hit submit too quickly.

> You mentioned that it was "more noticeable" without saying what that was.
Sorry, that was my confusion. What I was trying to say in the first post was that I didn't think you needed the Customer object (I stand corrected now) and that I fetched the customer object when it was one id when I probably didn't need to. Then I saw you say "more noticeable" and skimmed my original message and thought "more noticeable from the users side".

As far as I see, no evidence of database slowness.